src/server/api.php

```diff
@@ -1009,17 +1024,69 @@ function meta_delphi() {
       }
     }
   } else if($source === 'sensors') {
-    if(require_all($data, array('auth', 'names', 'locations', 'epiweeks'))) {
-      if($_REQUEST['auth'] === $AUTH['sensors']) {
-        // parse the request
-        $names = extract_values($_REQUEST['names'], 'str');
-        $locations = extract_values($_REQUEST['locations'], 'str');
-        $epiweeks = extract_values($_REQUEST['epiweeks'], 'int');
-        // get the data
-        $epidata = get_sensors($names, $locations, $epiweeks);
-        store_result($data, $epidata);
+    if(require_all($data, array('names', 'locations', 'epiweeks'))) {
+      if(!array_key_exists('auth', $_REQUEST)) {
+        $auth_tokens_presented = array();
       } else {
-        $data['message'] = 'unauthenticated';
+        $auth_tokens_presented = extract_values($_REQUEST['auth'], 'str');
+      }
+      $names = extract_values($_REQUEST['names'], 'str');
+      $n_names = count($names);
+      $n_auth_tokens_presented = count($auth_tokens_presented);
+      $max_valid_granular_tokens_per_name = max(array_map("count", $GRANULAR_SENSOR_AUTH_TOKENS));
+      // The number of valid granular tokens is related to the number of auth token checks that a single query could perform.  Use the max number of valid granular auth tokens per name in the check below as a way to prevent leakage of sensor names via this interface.  Treat all sensors as non-open for convenience of calculation.
+      if($n_names === 0) {
+        // Check whether no names were provided to prevent edge-case issues if surrounding behavior changes in the future:
+        $data['message'] = 'no sensor names provided';
+      } else if($n_auth_tokens_presented !== 1) {
```

Comment:
In this case, no auth tokens given, it looks like open sensors won't be served. Am I reading that correctly? If so, is that the behavior you want? I thought you wanted to return open sensors even when an auth token is missing.

Reply:
Nice catch!  I was trying to prevent a potential (future) situation where we might return all sensors by default, but broke the open sensor functionality instead.  Some unit tests can guard against that scenario instead.

Reply:
Fixed this and added some tests.  (Plus fixed naming: quidel->quid, and added missing wiki sensor key.)